\section{Promises}

Javascript/Typescript ist grundsätzlich eine sequenziell ausführende Skriptsprache. Das heißt,  dass zwei Bits eines Skripts nicht im gleichen Augenblick ausführbar sind. Sie müssen nacheinander ausgeführt werden. Im Browser können neben Javascript, auch CSS aktualisiert oder der Einfluss von außen, wie User-Interaktionen, ausgeführt werden. Und jede Operation verlangsamt die Andere. Um dem entgegenzuwirken hat man in dieser Arbeit schon das Prinzip der Callbacks gesehen. Dank der Rückruffunktionen ist die Ausführung dieser Sprache in asynchron möglich. Mit Callbacks kann man beispielsweise User-Events gut steuern. Ein User drückt auf die Enter-Taste und nach dem Event wird eine Funktionalität aufgerufen. Diese können in multiplen Ausführungen geschehen. Sollten jedoch Events von anderen Ressourcen abhängen, wie z.B. das Event darf erst eine Funktionalität \glqq{}feuern\grqq{} wenn das Bild im Browser geladen ist. Und sollte das Bild nicht geladen sein, sollte eine entsprechende Fehlermeldung angezeigt werden. Der Code hierfür würde wie folgt aussehen:

\begin{figure}[H]
\begin{lstlisting}
img1.callThisIfLoadedOrWhenLoaded(/** Callback 1 **/ function() {
  // Add Eventlistener to change the background
}).orIfFailedCallThis(/** Callback 2 **/ function() {
  // Show error msg for the user
});
\end{lstlisting}
\end{figure}

\noindent
Und genau das bieten \textbf{Promises} in einer übersichtlichen Form.

\begin{figure}[H]
\begin{lstlisting}
img1.ready().then(function() {
  // Add Eventlistener to change the background
}).catch(function() {
  // Show error msg for the user
});
\end{lstlisting}
\caption{Beispiel inspiriert von Google-Developer-Docs \cite{callback-vs-promises}}
\end{figure}

\noindent
Im Gegensatz zu Callbacks sind Promises an einer einzelner asynchronen erfolgreichen/fehlgeschlagenen Interaktion interessiert und weniger wann die Interaktion stattgefunden hat. Promises sind danach gerichtet wie man mit Ergebnis der Interaktion umgeht. 
Doch was genau sind Promises? \\

\subsection{Funktionsweise}

\noindent
Sie \textit{(\glqq{}Promises\grqq{} zu deutsch: Versprechen)} in Javascript verhalten sich ähnlich wie im echten Leben. Die Definition aus dem Wörterbuch ist: Das Versprechen ist eine einseitige Zusage über eine zukünftige Handlung oder ein zukünftiges Ereignis. \cite{versprechen} \\

\noindent
Das heißt:

\begin{enumerate}
    \item Ein Versprechen ist eine Absicherung, dass etwas gemacht wird. Unabhängig davon, ob das Versprechen sich selbst oder von einer anderen Partei gegeben wird.
    
    \item Ein Versprechen kann eingehalten oder gebrochen werden.
    
    \item Wurde ein Versprechen nicht eingehalten, möchte man den Grund für die Nichteinhaltung wissen um darauffolgend zu handeln.
    
    \item Beim Zeitpunkt eines Versprechens hat man nur die Absicherung. Man kann damit erstmal noch nichts anfangen. Es kann nur geplant werden was nach dem Einhalten des Versprechens gemacht wird. Dementsprechend kann man auch Maßnahmen setzen beim Nichteinhalten dieser Absicherung.
    
\end{enumerate}

\noindent
Und genauso ist das Verhalten der Promises in Javascript. Dabei gibt es zwei grundlegende Prinzipien der Promises, die zu Verstehen sind: Das \textbf{Erstellen von Promises} und das \textbf{Verarbeiten von Promises}.

\subsubsection{Erstellen eines Promises}

\begin{figure}[H]
\begin{lstlisting}
new Promise( /* executor */ function(resolve, reject) { ... } );
\end{lstlisting}
\caption{Erzeugung einer neuen Promise-Instanz}
\end{figure}

Der Konstruktor nimmt eine Rückruffunktion als Eingangsparameter. Diese Funktion wird auch \textbf{Executor} genannt.\cite{promise-executor} Der Executor akzeptiert zwei Parameter \textbf{resolve} und \textbf{reject}. Innerhalb dieser Funktion wird eine asynchrone Operation initiiert (z.B. Das suchen einer Datei, eine Datenbankabfrage etc.). Wurde diese asynchrone Operation erfolgreich ausgeführt, ruft der Promise-Konstruktor die resolve Funktion mit dem entsprechendem Ergebnis auf. Anders, bei einem unerwartetem Fehler, ruft der Konstruktor die reject Funktion mit der jeweiligen Fehlernachricht auf. Zur Einführung ein einfaches Beispiel:\\

\noindent
Vor dem Ausführen des Beispiels muss folgend konfiguriert werden:

 \begin{center}
     Promises-vs.-Observables$\,\to\,$ webpack.config.js
 \end{center}

\begin{figure}[H]
\begin{lstlisting}
module.exports = {
    mode: 'development',
    entry: './src/modules/modules/introduction.ts',
    ...
}
\end{lstlisting}
\end{figure}

\begin{figure}[H]
\begin{lstlisting}
const keepsHisWord = true;
const first = new Promise(function(resolve, reject) {
    if (keepsHisWord) {
        resolve('Promises kept!');
    } else {
        reject('Promise NOT kept!');
    }
});

console.log(first);
\end{lstlisting}
\end{figure}

\noindent
Dieser Promise löst sich auf Anhieb auf und der Status gelangt in den resolved Zustand, da die abhängige boolean-Variable vorher auf true gesetzt wurde. Umgekehrt, würde die boolean Variable auf false gesetzt werden, würde der Status rejected entsprechen. Der Initialstatus eines Promise wird im nächsten Beispiel verdeutlicht:

\begin{figure}[H]
\centering
\includegraphics{promise-beispiel-1}
\caption{Promises haben einen Status und einen Wert}
\end{figure}

\begin{figure}[H]
\begin{lstlisting}
export interface FakeHttpResponse {
    code: string;
    message: string;
}

const second = new Promise<FakeHttpResponse>(function(resolve, reject) {
    setTimeout(() => {
        resolve({
            code: '200',
            message: 'Promise kept!'
        });
    }, 10 * 1000);
});

console.log(second);
setTimeout(() => console.log(second), 10 * 1000);
\end{lstlisting}
\end{figure}

\noindent
Im oberen Beispiel wird der Promise vorbehaltlos nach zehn Sekunden aufgelöst, solange ist der Status ausstehend (pending). Nachdem der Promise aufgelöst wurde, werden Status und Wert aktualisiert. Dabei können nicht nur primitive Typen wie number, boolean oder string zurückgegeben werden, sondern auch generierte und komplexe Typen. Mit anderen Worten: Promises sind generisch.

\begin{figure}[H]
\centering
\includegraphics{promise-beispiel-2}
\caption{Promise hat anfangs den Status \glqq{}ausstehend\grqq{}}
\end{figure}

\noindent
Wie man nun gesehen hat kann der Promise-Status entweder auf \textbf{pending}, \textbf{resolved} oder \textbf{rejected} laufen. Im Status pending ist der Ausgang der Aktion noch ungewiss und deshalb der Promise-Wert undefined. Ändert sich der Status bzw. ist die Aktion zum Promise eingeschlagen oder fehlgeschlagen, ist der Status \textbf{settled}. Grundsätzlich läuft ein Promise also vom pending in den settled Status. In der nächsten Sektion wird auf das Verarbeiten von Promises näher eingegangen.

\subsubsection{Verarbeiten von Promises}

Nochmal zur Wiederholung: Ein Promise-Objekt repräsentiert das eventuelle Einschlagen oder Fehlschlagen einer asynchronen Operation, einschließlich des eingetroffenen Wertes. \\

\noindent
Ein solches Objekt bietet \textbf{statische} Methoden und \textbf{Prototyp}-Methoden. Die statischen Methoden können unabhängig von der Instanz aufgerufen werden, während die Prototyp-Methoden nur mit einer Instanz eines Promise-Objekts aufgerufen werden können. Es gibt drei Prototyp-Methoden. Alle der folgenden Methoden lassen sich unter einem Promise-Event einordnen.
Zur Wiederholung der verschiedenen Promise-Events:

\begin{itemize} 
\item Pending: Der Ausgang des Promises ist noch(!) ungewiss.
\item Resolved: Die Aktion, die zum Promise verlief, schlug ein.
\item Rejected: Die Aktion, die zum Promise verlief, schlug fehl.
\item Settled: Die Aktion ist entweder fehlgeschlagen oder eingeschlagen - jedoch abgeschlossen.
\end{itemize}

\noindent
Eine oder mehrere der drei Prototyp-Methoden werden aufgerufen wenn ein Promise in den settled Zustand übergelaufen ist:

\begin{itemize}

\begin{figure}[H]
\item \begin{lstlisting}
Promise.prototype.catch(onRejected)
\end{lstlisting}
\end{figure}

\begin{figure}[H]
\item \begin{lstlisting}
Promise.prototype.then(onFulfilled, onRejected)
\end{lstlisting}
\end{figure}
 
\begin{figure}[H]
\item \begin{lstlisting} 
Promise.prototype.finally(onFinally)
\end{lstlisting}
\end{figure}
 
\end{itemize}

\noindent
Die untenstehende Grafik stellt den Ablauf für die \textbf{then} und \textbf{catch} Methoden. Da beide Methoden ein Promise-Objekt zurückgeben, können Promises reibungslos aneinandergekettet werden. Wenn \textbf{finally} an ein Promise Objekt angebunden wird, wird diese Callback-Funktion in jedem Fall aufgerufen, wenn der Promise in den settled Zustand eingetreten ist, unabhängig davon ob der Promise \textbf{eingeschlagen} oder \textbf{fehlgeschlagen} ist.


\begin{figure}[H]
\includegraphics[width=12cm, height=6cm]{Promises-workflow}
\caption{Ablauf eines Promise-Operation \cite{promise-executor}}
\end{figure}

\noindent
Wie in der Grafik zusehen ist kann die Methode then 2 Argumente als Parameter entgegennehmen - einen für das einschlagen der Operation und einen für das fehlschlagen.

Ein Beispiel dafür wäre:

\begin{figure}[H]
\begin{lstlisting}
get('story.json').then(function(response) {
  console.log("Success!", response);
}, function(error) {
  console.log("Failed!", error);
})
\end{lstlisting}
\caption{Error-Handling ohne catch. \cite{callback-vs-promises}}
\end{figure}

\begin{figure}[H]
\begin{lstlisting}
get('story.json').then(function(response) {
  console.log("Success!", response);
}).catch(function(error) {
  console.log("Failed!", error);
})
\end{lstlisting}
\caption{Error-Handling mit catch. \cite{callback-vs-promises}}
\end{figure}

\noindent
Die catch Methode führt dabei keine zusätzlichen Operationen aus als die obere Variante. Sie macht den Code lediglich lesbarer. Wichtig zu beachten ist, dass innerhalb einer Executor-Funktion \textbf{niemals} beide Argumente gemeinsam eintreffen können. Sie verhalten sich exklusiv. Deshalb ist das letztere Beispiel im Verhalten vergleichbar mit diesem Beispiel:

\begin{figure}[H]
\begin{lstlisting}
get('story.json').then(function(response) {
  console.log("Success!", response);
}).then(undefined, function(error) {
  console.log("Failed!", error);
})
\end{lstlisting}
\end{figure}
\subsection{Operatoren}

Der Unterschied ist minimal, aber extrem hilfreich. Da Promise-Fehlschläge zu der nächsten then Methode in den fehlgeschlagen Callback weiterlaufen (oder in die catch Methode, wenn vorhanden). Mit then(func1, func2) werden entweder die erste Funktion oder die zweite Funktion aufgerufen, aber niemals beide. Jedoch mit then(func1).catch(func2) werden beide Callbacks aufgerufen wenn die erste Funktion fehlschlägt. Dies ist nur Möglich da, die Funktionen in unterschiedlichen Stellen der Verkettung liegen.


\begin{enumerate} 
\item Promise.all()
\item Promise.race() 
\item Promise.reject()
\end{enumerate}

\subsection{Promise Verkettung}
**Anwendungsbeispiel anzeigen**

\subsection{Async await}
**Funktionsweise und Vorteile**

\subsubsection{Anwendungsfall}

\begin{enumerate} 
\item Source code Beispiel zeigen mit async await
\item Wie wandle ich z.b. eine verkettete (unlesbare) Promise-Verkettung in async await um 
\item Gegenüberstellung
\item Browser kompatibilität
\end{enumerate}


